library/src/main/java/com/schibsted/spain/barista/action/BaristaListActions.kt

```diff
@@ -0,0 +1,109 @@
+import com.schibsted.spain.barista.internal.viewaction.PerformClickAction.clickUsingPerformClick
+import org.hamcrest.Matchers.*
+
+object BaristaListActions {
+
+    @JvmStatic
+    @JvmOverloads
+    fun clickListItem(@IdRes id: Int? = null, position: Int) {
+        performMagicAction(id, position,
+                recyclerAction = actionOnItemAtPosition<ViewHolder>(position, clickUsingPerformClick()),
+                listViewAction = clickUsingPerformClick()
+        )
+    }
+
+    @JvmStatic
+    @JvmOverloads
+    fun scrollListToPosition(@IdRes id: Int? = null, position: Int) {
+        performMagicAction(id, position,
+                recyclerAction = scrollToPosition<ViewHolder>(position),
+                listViewAction = scrollTo()
+        )
+    }
+
+    @JvmStatic
+    @JvmOverloads
+    fun clickListItemChild(@IdRes id: Int? = null, position: Int, @IdRes childId: Int) {
+        performMagicAction(id, position,
+                recyclerAction = actionOnItemAtPosition<ViewHolder>(position, clickChildWithId(childId)),
+                listViewAction = clickChildWithId(childId)
+        )
+    }
+
+    private fun performMagicAction(@IdRes id: Int?, position: Int, recyclerAction: ViewAction, listViewAction: ViewAction) {
+        verifyOneSingleMatch(id)
+
+        val spyFailureHandler = SpyFailureHandler()
+        try {
+            performOnRecycler(id, spyFailureHandler, recyclerAction)
+        } catch (noRecyclerMatching: NoMatchingViewException) {
+            performOnListView(spyFailureHandler, id, position, listViewAction)
+        }
+    }
+
+    private fun verifyOneSingleMatch(@IdRes id: Int?) {
+        val spyFailureHandler = SpyFailureHandler()
+        try {
+            onView(anyOf(findListViewMatcher(id), findRecyclerMatcher(id)))
+                    .withFailureHandler(spyFailureHandler)
+                    .check(matches(isDisplayed()))
+        } catch (multipleViews: AmbiguousViewMatcherException) {
+            spyFailureHandler.resendLastError("There are multiple ListView or RecyclerView in the hierarchy. You must specify an id parameter using clickListItem(id, position)")
+        } catch (noneFound: NoMatchingViewException) {
+            val message = if (id == null) {
+                "No ListView or RecyclerView found in the hierarchy"
+            } else {
+                "No ListView or RecyclerView with id ${resourceName(id)} was found in the hierarchy. Did you use a wrong id?"
+            }
+            spyFailureHandler.resendLastError(message)
+        }
+    }
+
+    private fun performOnRecycler(id: Int?, failureHandler: FailureHandler, recyclerAction: ViewAction) {
+        onView(findRecyclerMatcher(id))
+                .withFailureHandler(failureHandler)
+                .perform(recyclerAction)
+    }
+
+    private fun performOnListView(failureHandler: FailureHandler, id: Int?, position: Int, listViewAction: ViewAction) {
+        withFailureHandler(failureHandler) {
```

Comment:
Unrelated to this PR, but reading this wanted to comment:

What about changing:
`withFailureHandler(failureHandler) { ... }` to `failureHandler { ... }` or `failureHandler.apply { ... }` or `failureHandler.using { ... }`

So instead of a global function: `inline fun withFailureHandler(failureHandler: FailureHandler, scope: () -> Unit) {}` it would be `operator inline fun FailureHandler.apply(scope: () -> Unit) {}` (for example)

The idea behind that change is that you are repeating "failureHandler" twice when writing it

Reply:
Sorry @sch-cbavel, I think you lost me.
I'm not seeing the difference between both approaches (apart from the syntax)

Reply:
I meant that with this approach, since you are usually using **failureHandler** as the name of the variable, you are writing the name of the class twice (except for the case) :)

with**FailureHandler**(**failureHandler**) {
-->
**failureHandler**.apply {

Sure it is not a big deal, but reduces verbosity and its DRYer

Reply:
Hmm aren't `apply` and `with` type of methods usually intended for doing operations over the object itself?
In this case we're not using the failureHandler inside the code block, we're just using it internally to workaround Espresso.

IMO it might confuse readers. Not sure if it's just me.

Reply:
Sorry. My fault. Wanted to put invoke, not apply:

```
class FailureHandler {
    operator fun invoke(callback: () -> Unit) {
        // use...
        try {
            callback()
        } finally {
            // restore
        }
    }
}

fun test() {
    val failureHandler = FailureHandler()

    failureHandler {
    }
    
    // or

    failureHandler.invoke {
    }
}
```

Other possible verbs:
```
failureHandler.temporally {
}
failureHandler.locally {
}
...
```

I'm pretty used to Kotlin DSLs and find this clean. But if you think it can confuse people. It is ok to leave as it is :) :+1: